mirror-main.go

```diff
@@ -245,221 +266,422 @@ func doMirror(sURLs mirrorURLs, progressReader *progressBar, accountingReader *a
+					continue
+				// Handle these specifically for object storage related errors.
+				case BucketNameEmpty, ObjectMissing, ObjectAlreadyExists, ObjectAlreadyExistsAsDirectory, BucketDoesNotExist, BucketInvalid, ObjectOnGlacier:
+					continue
 				}
+
+				// For critical errors we should exit. Session
+				// can be resumed after the user figures out
+				// the  problem. We know that
+				// there are no current copy actions, because
+				// we're not multi threading now.
+
+				// this issue could be separated using separate
+				// error channel instead of using sURLs.Error
+				ms.CloseAndDie()
+			}
+
+			// finished harvesting urls, save queue to session data
+			if err := ms.queue.Save(ms.NewDataWriter()); err != nil {
+				ms.status.fatalIf(probe.NewError(err), "Unable to save queue.")
 			}
-			// Remember totalBytes only for mirror not for removal,
+
 			if sURLs.SourceContent != nil {
-				totalBytes += sURLs.SourceContent.Size
+				ms.Header.LastCopied = sURLs.SourceContent.URL.String()
+			} else if sURLs.TargetContent != nil {
+				ms.Header.LastRemoved = sURLs.TargetContent.URL.String()
 			}
-			totalObjects++
-		case <-trapCh:
-			// Print in new line and adjust to top so that we don't print over the ongoing scan bar
-			if !globalQuiet && !globalJSON {
-				console.Eraseline()
+
+			ms.Save()
+
+			if sURLs.SourceContent != nil {
+			} else if sURLs.TargetContent != nil {
+				// Construct user facing message and path.
+				targetPath := filepath.ToSlash(filepath.Join(sURLs.TargetAlias, sURLs.TargetContent.URL.Path))
+				ms.status.PrintMsg(rmMessage{
+					Status: "success",
+					URL:    targetPath,
+				})
 			}
-			session.Delete() // If we are interrupted during the URL scanning, we drop the session.
-			os.Exit(0)
 		}
-	}
-	session.Header.TotalBytes = totalBytes
-	session.Header.TotalObjects = totalObjects
-	session.Save()
+	}()
 }
 
-// Session'fied mirror command.
-func doMirrorSession(session *sessionV7) {
-	isForce := session.Header.CommandBoolFlags["force"]
-	isFake := session.Header.CommandBoolFlags["fake"]
-	isRemove := session.Header.CommandBoolFlags["remove"]
+func (ms *mirrorSession) startMirror(wait bool) {
+	isRemove := ms.Header.CommandBoolFlags["remove"]
 
-	// Initialize signal trap.
-	trapCh := signalTrap(os.Interrupt, syscall.SIGTERM)
+	ms.wgCopy.Add(1)
 
-	if !session.HasData() {
-		doPrepareMirrorURLs(session, isForce, isFake, isRemove, trapCh)
-	}
+	go func() {
+		defer ms.wgCopy.Done()
 
-	// Enable accounting reader by default.
-	accntReader := newAccounter(session.Header.TotalBytes)
+		for {
+			if !wait {
+			} else if err := ms.queue.Wait(); err != nil {
+				break
+			}
 
-	// Set up progress bar.
-	var progressReader *progressBar
-	if !globalQuiet && !globalJSON {
-		progressReader = newProgressBar(session.Header.TotalBytes)
-	}
+			v := ms.queue.Pop()
+			if v == nil {
+				break
+			}
+
+			sURLs := v.(mirrorURLs)
```

Comment:
Needs proper type assertion here, with 'ok' better we avoid non-typed returns in the first place.